cpp/src/io/utilities/datasource.cpp

```diff
@@ -199,6 +200,65 @@ class direct_read_source : public file_source {
   }
 };
 
+/**
+ * @brief Implementation class for reading from a device buffer source
+ */
+class device_buffer_source final : public datasource {
+ public:
+  explicit device_buffer_source(cudf::device_span<std::byte const> d_buffer) : _d_buffer{d_buffer}
+  {
+  }
+
+  size_t host_read(size_t offset, size_t size, uint8_t* dst) override
+  {
+    auto const count = std::min(size, this->size() - offset);
+    CUDF_CUDA_TRY(cudaMemcpy(dst, _d_buffer.data() + offset, count, cudaMemcpyDefault));
```

Comment:
Should this use `cudaMemcpyAsync` with a stream? (Even if it must be synced immediately -- it might mean that it's only a stream sync instead of a device sync.)

See previous conversation with @vuule: https://github.com/rapidsai/cudf/pull/12314#discussion_r1060934999

> I default to `cudaMemcpyAsync` because it takes a stream, and we seem to be completely avoiding `cudaMemcpy` in libcudf (presumably because of the use of default stream?).

Reply:
> Should this use cudaMemcpyAsync with a stream? 

Yes, this has been discussed in https://github.com/rapidsai/cudf/pull/12024#discussion_r1066192873. Adding the stream parameter requires changes to all classes derived from the base `datasource` thus would be done in a separate PR.

Reply:
I'm suggesting something different that doesn't require a new parameter. Replace:
```cpp
CUDF_CUDA_TRY(cudaMemcpy(...));
```
with
```cpp
auto const stream = cudf::get_default_stream();
CUDF_CUDA_TRY(cudaMemcpyAsync(..., stream);
stream.synchronize();
```

From my understanding, `cudaMemcpy` syncs the whole device and not just the stream that libcudf is using. If that's incorrect, please let me know!

Reply:
Got it. Make sense to me.